### PR TITLE
[utils/bloom] Dynamic Bloom Filter

### PR DESCRIPTION
**Pull request by patrick-ogrady**
## Why this should be merged

Rationale: https://github.com/ava-labs/avalanchego/pull/2591

## How this works

Modifies reset logic to grow the bloom filter if the number of target items grows.

## How this was tested

CI + Load Test

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,7 +4,7 @@ go 1.20
 
 require (
 	github.com/VictoriaMetrics/fastcache v1.10.0
-	github.com/ava-labs/avalanchego v1.10.18-rc.15
+	github.com/ava-labs/avalanchego v1.10.18-rc.5.0.20240105233604-41587de9c530
```
Comment (patrick-ogrady) [inline go.mod:7]: will update before merging

File: plugin/evm/vm.go [hunk 2]
```diff
@@ -141,15 +141,15 @@ const (
...
-	txGossipBloomMaxItems          = 8 * 1024
-	txGossipBloomFalsePositiveRate = 0.01
-	txGossipMaxFalsePositiveRate   = 0.05
-	txGossipTargetMessageSize      = 20 * units.KiB
-	maxValidatorSetStaleness       = time.Minute
-	txGossipThrottlingPeriod       = 10 * time.Second
-	txGossipThrottlingLimit        = 2
-	gossipFrequency                = 10 * time.Second
-	txGossipPollSize               = 10
+	txGossipBloomMinTargetItems          = 1024
```
Comment (patrick-ogrady) [inline plugin/evm/vm.go:144]: now that this scales up automatically, we don't need to set such a high minimum

File: plugin/evm/gossip.go [hunk 3]
```diff
@@ -139,10 +139,11 @@ func (g *GossipEthTxPool) Subscribe(ctx context.Context) {
 			return
 		case pendingTxs := <-g.pendingTxs:
 			g.lock.Lock()
+			optimalSize := g.mempool.PendingSize() + len(pendingTxs.Txs) // should never come close to overflowing
```
Comment (patrick-ogrady) [inline plugin/evm/gossip.go:142]: consider adding a churn multiplier to this to avoid frequent refresh

File: plugin/evm/vm.go (hunk 2 above)
Comment (patrick-ogrady) [reply plugin/evm/vm.go:144]: We may want to change the name to `txGossipBloomMinTargetElements`

File: plugin/evm/vm.go [hunk 4]
```diff
@@ -141,15 +141,16 @@ const (
 	atomicTxGossipProtocol = 0x1
 
 	// gossip constants
-	txGossipBloomMaxItems          = 8 * 1024
```
Comment (patrick-ogrady) [inline plugin/evm/vm.go:144]: I think this change is worth a discussion (ok to leave as is)

Comment (StephenButtolph) [reply plugin/evm/vm.go:144]: I'd prefer to leave it

Comment (patrick-ogrady) [reply plugin/evm/vm.go:144]: Updated

**Review by StephenButtolph — APPROVED** [review]

File: plugin/evm/vm.go (hunk 4 above)
Comment (ceyonur) [reply plugin/evm/vm.go:144]: would keeping this value same for new `OptimalParameters` mean that we expect 8*1024 entries in the filter?

Comment (patrick-ogrady) [reply plugin/evm/vm.go:144]: This actually does the same thing but the previous variable name was confusing (as you pointed out 😉).

**Review by ceyonur — APPROVED** [review]

**Pull request merged**